doc/horizontal_autoscaling.md

```diff
@@ -0,0 +1,258 @@
+        memory: "600Mi"
+```
+
+The training job controller will create and continuously scale the
+number of trainers and the number of pservers between the
+corresponding `min-instance` and `max-instance`.
+
+Since the `master` server is necessary only when the trainer is using
+`paddle.v2.reader.creator.cloud_reader`, The `master` spec is
+optional: the master server will be created only when configured.
+
+The training job custom resource can be created with: `kubectl create
+-f training_job.yaml`.
+
+The custom resource will only be saved on Kuberentes, we will need a
+custom controller that operate on it.
+
+### Training Job Controller
+
+The training job controller run as a Pod. It has the global view of
+the computation resources. It watches the training job resources and
+schedules and scales the training jobs using the Kuberenetes API.
+
+The pseudo controller declaration (`training_job_controller.yaml`) is
+as follows:
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: training-job-controller
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        name: training-job-controller
+    spec:
+      containers:
+      - name: training-job-controller
+        image: paddlepaddle/training-job-controller
+```
+
+The training job controller can be started by the cluster
+administrator with command: `kubectl create -f
+training_job_controller.yaml`
+
+## Implementation
+
+### Training Job Resource
+
+The training job resource is a custom resource, there are two ways of
+implementing custom resources:
+
+- [Custom Resource Definition (CRD)](https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/),
+  since Kuberentes v1.7.
+- [Third Party Resource (TPR)](https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-third-party-resource/),
+  since Kubernetes v1.2, fully deprecated in v1.8, will be removed in
+  v1.9.
+
+We will support TPR first, because some of our clusters is using
+Kubernetes v1.6. CRD will be supported in the future.
+
+
+### Training Job Controller
+
+Currently, we will run a single training job controller instance and
+assume that there is no training job controller running concurrently
+(the assumption could be false
+when
+[split-brain](https://en.wikipedia.org/wiki/Split-brain_(computing))
+happens). In the future, we will run multiple instances and use leader
+election to choose a leader.
+
+The pseudo-logic is as follows:
+
+```go
+registerThirdPartyResource()
+for {
+  quota := getTotalComputationResourceQuota()
+  current := getCurrentJobStates()
+  desired := getDesiredJobStates()
+  dynamicScaling(quota, current, desired)
+}
+```
+
+#### Scaling Algorithm
+
+##### Elastic Job
+
+A job is elastic only when it's trainer and pserver's `min-instance`
+equals to the `max-instance` respectively. We will only scale elastic
+jobs.
+
+Currently, we will not scale the parameter server instances.
+
+##### Fulfillment Score
```

Comment:
Fulfillment score is awesome. Maybe it's not enough just use it.
Do we need to classify the jobs by level and high-level tasks are prioritized?

Reply:
That's a good point, can we implement the current version first. And when we have real need for the task-priority-level, we can change this design doc and implement later?

Reply:
Agree with you.